File: src/test/java/com/rbmhtechnology/calliope/StreamExecutorSpec.java
/*
 * Copyright 2015 - 2017 Red Bull Media House GmbH <http://www.redbullmediahouse.com> - all rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.rbmhtechnology.calliope;

import akka.Done;
import akka.actor.ActorRef;
import akka.actor.ActorSystem;
import akka.stream.javadsl.Keep;
import akka.stream.javadsl.RunnableGraph;
import akka.stream.javadsl.Sink;
import akka.stream.javadsl.Source;
import akka.testkit.JavaTestKit;
import akka.testkit.TestProbe;
import com.rbmhtechnology.calliope.StreamExecutor.RunStream;
import com.rbmhtechnology.calliope.StreamExecutor.StreamStarted;
import io.vavr.collection.Seq;
import io.vavr.collection.Vector;
import org.junit.ClassRule;
import org.junit.Test;
import org.reactivestreams.Publisher;
import org.reactivestreams.Subscriber;
import org.scalatest.junit.JUnitSuite;

import java.time.Duration;
import java.util.concurrent.CompletionStage;

import static com.rbmhtechnology.calliope.StreamExecutorSupervision.Directive.STOP;
import static com.rbmhtechnology.calliope.javadsl.Durations.toFiniteDuration;

public class StreamExecutorSpec extends JUnitSuite {

  @ClassRule
  public static AkkaStreamsRule akka = new AkkaStreamsRule("reference.conf");

  private final ReplayProducer<String> producer = ReplayProducer.create();

  private final StreamExecutorSupervision.Backoff supervision =
    StreamExecutorSupervision.withBackoff(Duration.ofMillis(50), Duration.ofMillis(100), 0.2);

  @Test
  public void executor_when_run_must_startStream() {
    new JavaTestKit(akka.system()) {{
      final TestProbe consumer = new TestProbe(getSystem());
      final ActorRef executor = getSystem().actorOf(StreamExecutor.props(actorConsumerGraph(producer, consumer.ref()), supervision));

      runStream(getSystem(), executor);

      producer.sendNext("event-1");
      consumer.expectMsg("event-1");
    }};
  }

  @Test
  public void executor_when_streamCompletesWithFailure_must_restartStream() throws InterruptedException {
    new JavaTestKit(akka.system()) {{
      final TestProbe consumer = new TestProbe(getSystem());
      final ActorRef executor = getSystem().actorOf(StreamExecutor.props(actorConsumerGraph(producer, consumer.ref()), supervision));

      runStream(getSystem(), executor);

      producer.sendNext("event-1");
      consumer.expectMsg("event-1");

      producer.sendError(new RuntimeException("err"));

      producer.sendNext("event-2");
      consumer.expectMsg("event-1");
      consumer.expectMsg("event-2");
    }};
  }

  @Test
  public void executor_when_streamCompletesSuccessfully_must_restartStream() throws InterruptedException {
    new JavaTestKit(akka.system()) {{
      final TestProbe consumer = new TestProbe(getSystem());
      final ActorRef executor = getSystem().actorOf(StreamExecutor.props(actorConsumerGraph(producer, consumer.ref()), supervision));

      runStream(getSystem(), executor);

      producer.sendNext("event-1");
      consumer.expectMsg("event-1");

      producer.sendComplete();

      producer.sendNext("event-2");
      consumer.expectMsg("event-1");
      consumer.expectMsg("event-2");
    }};
  }

  @Test
  public void executor_when_stoppingDirectiveSelectedBySupervisionDecider_must_stopStream() throws InterruptedException {
    new JavaTestKit(akka.system()) {{
      final TestProbe consumer = new TestProbe(getSystem());
      final StreamExecutorSupervision stoppingSupervision = supervision.withDecider(err -> STOP);

      final ActorRef executor = getSystem()
        .actorOf(StreamExecutor.props(actorConsumerGraph(producer, consumer.ref()), stoppingSupervision));

      runStream(getSystem(), executor);

      producer.sendNext("event-1");
      consumer.expectMsg("event-1");

      producer.sendError(new RuntimeException("err"));

      producer.sendNext("event-2");
      consumer.expectNoMsg(toFiniteDuration(Duration.ofSeconds(1)));
    }};
  }

  private <A> RunnableGraph<CompletionStage<Done>> actorConsumerGraph(final Publisher<A> producer, final ActorRef consumer) {
    return Source.fromPublisher(producer)
      .map(em -> {
        consumer.tell(em, ActorRef.noSender());
        return em;
      })
      .toMat(Sink.ignore(), Keep.right());
  }

  private ActorRef runStream(final ActorSystem system, final ActorRef streamExecutor) {
    final TestProbe complete = new TestProbe(system);
    streamExecutor.tell(RunStream.instance(), complete.ref());
    complete.expectMsg(StreamStarted.instance());
    return streamExecutor;
  }

  private static class ReplayProducer<A> implements Publisher<A> {

    private Seq<Subscriber<? super A>> subscribers;
    private Seq<A> emissions;

    private ReplayProducer(final Seq<Subscriber<? super A>> subscribers, final Seq<A> emissions) {
      this.subscribers = subscribers;
      this.emissions = emissions;
    }

    private static <A> ReplayProducer<A> create() {
      return new ReplayProducer<>(Vector.empty(), Vector.empty());
    }

    @Override
    public void subscribe(final Subscriber<? super A> s) {
      subscribers = subscribers.append(s);
      emissions.forEach(s::onNext);
    }

    public void sendNext(final A element) {
      emissions = emissions.append(element);
      subscribers.forEach(s -> s.onNext(element));
    }

    public void sendError(final Throwable error) {
      subscribers.forEach(s -> s.onError(error));
    }

    public void sendComplete() {
      subscribers.forEach(Subscriber::onComplete);
    }
  }
}

File: src/test/scala/com/rbmhtechnology/calliope/StreamExecutorSpec.scala
/*
 * Copyright 2015 - 2017 Red Bull Media House GmbH <http://www.redbullmediahouse.com> - all rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.rbmhtechnology.calliope

import java.time.Duration
import java.util.concurrent.CompletionStage

import akka.{Done, NotUsed}
import akka.actor.{ActorRef, ActorSystem}
import akka.stream.javadsl.{Keep, RunnableGraph, Sink}
import akka.stream.scaladsl.Source
import akka.testkit.{TestKit, TestProbe}
import com.rbmhtechnology.calliope.StreamExecutor.{RunStream, StreamStarted}
import com.rbmhtechnology.calliope.StreamExecutorSupervision.Decider
import com.rbmhtechnology.calliope.StreamExecutorSupervision.Directive.STOP
import org.reactivestreams.{Publisher, Subscriber}
import org.scalatest.{BeforeAndAfterEach, MustMatchers, WordSpecLike}

import scala.collection.immutable
import scala.concurrent.duration._

object StreamExecutorSpec {

  case class ReplayProducer[A]() extends Publisher[A] {

    private var subscribers: immutable.Seq[Subscriber[_ >: A]] = Vector.empty
    private var emissions: immutable.Seq[A] = Vector.empty

    override def subscribe(s: Subscriber[_ >: A]): Unit = {
      subscribers = subscribers :+ s
      emissions.foreach(s.onNext(_))
    }

    def sendNext(element: A): Unit = {
      emissions = emissions :+ element
      subscribers.foreach(_.onNext(element))
    }

    def sendError(err: Throwable): Unit = {
      subscribers.foreach(_.onError(err))
    }

    def sendComplete(): Unit = {
      subscribers.foreach(_.onComplete())
    }
  }
}

class StreamExecutorSpec extends TestKit(ActorSystem("test")) with WordSpecLike with MustMatchers with BeforeAndAfterEach with StreamSpec {

  import StreamExecutorSpec._

  private val supervision = StreamExecutorSupervision.withBackoff(Duration.ofMillis(50), Duration.ofMillis(100), 0.2)

  var consumer: TestProbe = _
  var producer: ReplayProducer[String] = _

  override protected def beforeEach(): Unit = {
    super.beforeEach()

    consumer = TestProbe()
    producer = ReplayProducer[String]()
  }

  def streamExecutor[A](producer: Publisher[A], consumer: TestProbe, supervision: StreamExecutorSupervision = supervision): ActorRef =
    system.actorOf(StreamExecutor.props(actorConsumerGraph(producer, consumer.ref), supervision))

  def actorConsumerGraph[A](producer: Publisher[A], consumer: ActorRef): RunnableGraph[CompletionStage[Done]] =
    Source.fromPublisher(producer)
      .map { m =>
        consumer.tell(m, ActorRef.noSender)
        m
      }
      .asJava
      .toMat(Sink.ignore, Keep.right[NotUsed, CompletionStage[Done]])

  def runStream(streamExecutor: ActorRef): ActorRef = {
    val complete = TestProbe()
    streamExecutor.tell(RunStream.instance, complete.ref)
    complete.expectMsg(StreamStarted.instance)
    streamExecutor
  }

  "A StreamExecutor" when {
    "started" must {
      "run the underlying stream" in {
        runStream(streamExecutor(producer, consumer))

        producer.sendNext("event-1")
        consumer.expectMsg("event-1")
      }
    }
    "stream completes with failure" must {
      "restart the stream" in {
        runStream(streamExecutor(producer, consumer))

        producer.sendNext("event-1")
        consumer.expectMsg("event-1")

        producer.sendError(new RuntimeException("error"))

        producer.sendNext("event-2")
        consumer.expectMsg("event-1")
        consumer.expectMsg("event-2")
      }
    }
    "stream completes successfully" must {
      "restart the stream" in {
        runStream(streamExecutor(producer, consumer))

        producer.sendNext("event-1")
        consumer.expectMsg("event-1")

        producer.sendComplete()

        producer.sendNext("event-2")
        consumer.expectMsg("event-1")
        consumer.expectMsg("event-2")
      }
    }
    "stopping-directive selected by supervision-decider" must {
      "stop the stream" in {
        val stoppingSupervision = supervision.withDecider(new Decider {
          override def apply(t: Throwable): StreamExecutorSupervision.Directive = STOP
        })
        runStream(streamExecutor(producer, consumer, stoppingSupervision))

        producer.sendNext("event-1")
        consumer.expectMsg("event-1")

        producer.sendError(new RuntimeException("error"))

        producer.sendNext("event-2")
        consumer.expectNoMsg(1.second)
      }
    }
  }
}
